File: articles/lifespan-developmental-psychology-and-the-baltes-tradition/python/model_lifespan_baltes.py
#!/usr/bin/env python3
"""Model synthetic lifespan developmental psychology in the Baltes tradition."""

from __future__ import annotations

from pathlib import Path
import matplotlib.pyplot as plt
import pandas as pd
import statsmodels.formula.api as smf

ROOT = Path(__file__).resolve().parents[1]
DATA_PATH = ROOT / "data" / "lifespan_baltes_panel.csv"
OUTPUTS_DIR = ROOT / "outputs"
OUTPUTS_DIR.mkdir(parents=True, exist_ok=True)


def main() -> None:
    if not DATA_PATH.exists():
        raise FileNotFoundError("Run python/generate_lifespan_baltes_panel.py first.")

    panel = pd.read_csv(DATA_PATH).sort_values(["id", "time"]).reset_index(drop=True)
    panel["lag_score"] = panel.groupby("id")["development_score"].shift(1)
    model_data = panel.dropna(subset=["lag_score"]).copy()

    development_model = smf.ols(
        """
        development_score ~ lag_score + time + gains + losses +
        plasticity + current_support + current_comp + health_resource +
        historical_support + institutional_security + soc_index +
        plasticity:current_support + losses:compensation
        """,
        data=model_data,
    ).fit(cov_type="HC3")

    soc_model = smf.ols(
        """
        soc_index ~ time + selection + optimization + compensation +
        current_support + current_comp + health_resource +
        historical_support + institutional_security
        """,
        data=panel,
    ).fit(cov_type="HC3")

    with open(OUTPUTS_DIR / "python_lifespan_baltes_model_summary.txt", "w", encoding="utf-8") as f:
        f.write("DYNAMIC LIFESPAN DEVELOPMENT MODEL\n")
        f.write("=" * 80 + "\n")
        f.write(development_model.summary().as_text())
        f.write("\n\nSOC ADAPTATION MODEL\n")
        f.write("=" * 80 + "\n")
        f.write(soc_model.summary().as_text())

    trajectory = panel.groupby("time", as_index=False).agg(
        average_development=("development_score", "mean"),
        average_gains=("gains", "mean"),
        average_losses=("losses", "mean"),
        average_soc=("soc_index", "mean"),
        standard_error=("development_score", lambda x: x.std() / (len(x) ** 0.5)),
    )
    trajectory["lower"] = trajectory["average_development"] - 1.96 * trajectory["standard_error"]
    trajectory["upper"] = trajectory["average_development"] + 1.96 * trajectory["standard_error"]
    trajectory.to_csv(OUTPUTS_DIR / "python_lifespan_trajectory.csv", index=False)

    plt.figure(figsize=(8, 5))
    plt.plot(trajectory["time"], trajectory["average_development"], marker="o")
    plt.fill_between(trajectory["time"], trajectory["lower"], trajectory["upper"], alpha=0.15)
    plt.xlabel("Time")
    plt.ylabel("Average development score")
    plt.title("Synthetic Lifespan Development in the Baltes Tradition")
    plt.tight_layout()
    plt.savefig(OUTPUTS_DIR / "python_lifespan_trajectory.png", dpi=160)
    plt.close()

    soc_trajectory = panel.groupby("time", as_index=False).agg(
        average_selection=("selection", "mean"),
        average_optimization=("optimization", "mean"),
        average_compensation=("compensation", "mean"),
        average_soc=("soc_index", "mean"),
    )
    soc_trajectory.to_csv(OUTPUTS_DIR / "python_soc_trajectory.csv", index=False)

    plt.figure(figsize=(8, 5))
    plt.plot(soc_trajectory["time"], soc_trajectory["average_selection"], marker="o", label="selection")
    plt.plot(soc_trajectory["time"], soc_trajectory["average_optimization"], marker="o", label="optimization")
    plt.plot(soc_trajectory["time"], soc_trajectory["average_compensation"], marker="o", label="compensation")
    plt.xlabel("Time")
    plt.ylabel("Average SOC component")
    plt.title("Synthetic Selection, Optimization, and Compensation")
    plt.legend()
    plt.tight_layout()
    plt.savefig(OUTPUTS_DIR / "python_soc_trajectory.png", dpi=160)
    plt.close()

    profile_summary = panel.groupby("adaptation_profile", as_index=False).agg(
        people=("id", "nunique"),
        average_development=("development_score", "mean"),
        average_gains=("gains", "mean"),
        average_losses=("losses", "mean"),
        average_support=("current_support", "mean"),
        average_soc=("soc_index", "mean"),
    )
    profile_summary.to_csv(OUTPUTS_DIR / "python_lifespan_profiles.csv", index=False)

    print("Wrote Python lifespan Baltes model outputs.")


if __name__ == "__main__":
    main()

File: articles/lifespan-developmental-psychology-and-the-baltes-tradition/r/model_lifespan_baltes.R
#!/usr/bin/env Rscript

suppressPackageStartupMessages({
  library(dplyr)
  library(ggplot2)
  library(lme4)
})

args <- commandArgs(trailingOnly = FALSE)
file_arg <- "--file="
script_path <- normalizePath(sub(file_arg, "", args[grep(file_arg, args)]))
root <- normalizePath(file.path(dirname(script_path), ".."), mustWork = TRUE)

data_path <- file.path(root, "data", "lifespan_baltes_panel.csv")
outputs_dir <- file.path(root, "outputs")

if (!dir.exists(outputs_dir)) dir.create(outputs_dir, recursive = TRUE)
if (!file.exists(data_path)) stop("Run python/generate_lifespan_baltes_panel.py first.")

panel <- read.csv(data_path)

development_model <- lmer(
  development_score ~ time + gains + losses + plasticity +
    current_support + current_comp + health_resource +
    historical_support + institutional_security + soc_index +
    plasticity:current_support + losses:compensation +
    (1 + time | cohort_id/id),
  data = panel
)

soc_model <- lmer(
  soc_index ~ time + selection + optimization + compensation +
    current_support + current_comp + health_resource +
    historical_support + institutional_security +
    (1 + time | cohort_id/id),
  data = panel
)

capture.output(
  list(
    development_model = summary(development_model),
    soc_model = summary(soc_model)
  ),
  file = file.path(outputs_dir, "r_lifespan_baltes_model_summary.txt")
)

trajectory <- panel |>
  group_by(time) |>
  summarize(
    average_development = mean(development_score),
    average_gains = mean(gains),
    average_losses = mean(losses),
    average_soc = mean(soc_index),
    standard_error = sd(development_score) / sqrt(n()),
    lower = average_development - 1.96 * standard_error,
    upper = average_development + 1.96 * standard_error,
    .groups = "drop"
  )

write.csv(trajectory, file.path(outputs_dir, "r_lifespan_trajectory.csv"), row.names = FALSE)

ggsave(
  filename = file.path(outputs_dir, "r_lifespan_trajectory.png"),
  plot = ggplot(trajectory, aes(x = time, y = average_development)) +
    geom_line(linewidth = 1) +
    geom_ribbon(aes(ymin = lower, ymax = upper), alpha = 0.12) +
    labs(title = "Synthetic Lifespan Development in the Baltes Tradition", x = "Time", y = "Average development") +
    theme_minimal(),
  width = 8,
  height = 5,
  dpi = 160
)

soc <- panel |>
  group_by(time) |>
  summarize(
    average_selection = mean(selection),
    average_optimization = mean(optimization),
    average_compensation = mean(compensation),
    average_soc = mean(soc_index),
    .groups = "drop"
  )

write.csv(soc, file.path(outputs_dir, "r_soc_trajectory.csv"), row.names = FALSE)

ggsave(
  filename = file.path(outputs_dir, "r_soc_trajectory.png"),
  plot = ggplot(soc, aes(x = time)) +
    geom_line(aes(y = average_selection, linetype = "selection"), linewidth = 1) +
    geom_line(aes(y = average_optimization, linetype = "optimization"), linewidth = 1) +
    geom_line(aes(y = average_compensation, linetype = "compensation"), linewidth = 1) +
    labs(title = "Synthetic Selection, Optimization, and Compensation", x = "Time", y = "Average SOC component", linetype = "Component") +
    theme_minimal(),
  width = 8,
  height = 5,
  dpi = 160
)

message("Wrote R lifespan Baltes outputs.")
